### DISCUSSION TITLE
Thanos Compactor seems to be running in circles

### DISCUSSION THREAD

**Discussion by vincent-olivert-riera** [thanos-io/thanos#7272] [Questions & Answers] [answered]
I have a Thanos Compactor tracing an S3 bucket where **no** new blocks are uploaded from Sidecars.

What I am observing is that the number of `thanos_compact_todo_compactions` goes up and down. Given that no Sidecars are uploading blocks to that bucket, I was expecting that metric to always go down.

![Screenshot 2024-04-11 at 13-08-00 Explore - thanos-release - Grafana](https://github.com/thanos-io/thanos/assets/6151586/b7ea025e-664b-4f4d-af4a-7a25e2bcf8e4)

The Compactor is not halted. I check the `thanos_compact_halted` metric from time to time and its value is `0`.

The bucket has a very large quantity of level-1 blocks, a few (20-ish) level-2 blocks, and no level-3 or level-4 blocks.
I can see that level-2 blocks are slowly increasing, so the Compactor seems to be doing its job, although very slowly.

Is that a normal behavior?

**Comment by GiedriusS** [comment, marked as answer]
It will not always go down because the compactor/planner isn't so smart in Thanos. It first compacts small blocks into bigger ones and then after a few compactions like this, it generates another plan to generate those bigger blocks into even bigger ones, and so on. A smarter algorithm would just compact a lot of small blocks into one big block. That's when you would see the backlog only go down. Alas, that's not the case. At least that's how I understand it

**Reply by vincent-olivert-riera** [reply]
I see. Glad to know that behavior is expected. Thanks 👍
